torch/distributed/_shard/checkpoint/metadata.py

```diff
@@ -0,0 +1,79 @@
+import io
+import pickle
+from dataclasses import dataclass
+from typing import Dict, List, Tuple, Union
+
+import torch
+from torch.distributed._shard.sharded_tensor import (
+    ShardedTensor,
+    ShardedTensorMetadata,
+    ShardMetadata,
+)
+
+TENSOR_TYPE = Union[torch.Tensor, ShardedTensor]
+
+@dataclass
+class ShardStorageMetadata:
+    shard_metadata: ShardMetadata
+    # Unique identifier for this particular Shard
+    storage_key: str
+    # Length in bytes for this shard
+    length: int
+
+
+# Metadata for each param.
+@dataclass
+class ShardedTensorStorageMetadata:
+    # Metadata for the sharded tensor itself
+    tensor_metadata: ShardedTensorMetadata
+    # Storage info for each Shard. There's no ordering requirement for this list.
+    storage_metadata: List[ShardStorageMetadata]
+
+
+@dataclass
+class TensorStorageMetadata:
```

Comment:
Is this actually sufficient in terms of metadata that we need? Do we need something like [TensorProperties](https://github.com/pytorch/pytorch/blob/ba205a7eae357135f937e0501f7748d2dd953eba/torch/distributed/_shard/sharded_tensor/metadata.py#L15) too?

For example `ShardedTensorMetadata` has all of this information. It seems like it would be useful to just query a checkpoint and get information about the tensor's shape and everything else as well.

Reply:
Reusing TensorProperties would be ideal. But this is blocked on: https://github.com/pytorch/pytorch/pull/76679